### PR TITLE
Retry SSL operations on SSL_ERROR_WANT_READ/WRITE

### PR DESCRIPTION
**Pull request by vpodzime**
When the client gets SSL_ERROR_WANT_READ or SSL_ERROR_WANT_WRITE,
we should re-try instead of shutting down the whole session
because these errors are expected, as documented in the OpenSSL
documentation (and functions' man pages)

Ticket: CFE-3023
Changelog: Reduce SSL/TLS shutdowns on bad networks

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @olehermanse can review this?

**Comment by olehermanse** [comment]
This pull request **introduces 1 alert** when merging fea2f7c485e851ce30637542ff68dcb5b24e597c into 6dafc4082363c8902eb17e5f2c25945e0c776b21 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-12117f8cb06bcd5db13105322eb2173e675d875f)

**new alerts:**

* 1 for Variable used in its own initializer

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by vpodzime** [comment]
@cf-bottom jenkins with exotics, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2374/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2374/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2374/

**Comment by vpodzime** [comment]
Jenkins is green except for Solaris 11 which got lost in the clouds.

File: libcfnet/tls_client.c [hunk 1]
```diff
@@ -310,8 +311,35 @@ int TLSTry(ConnectionInfo *conn_info)
     /* Initiate the TLS handshake over the already open TCP socket. */
     SSL_set_fd(conn_info->ssl, conn_info->sd);
 
-    int ret = SSL_connect(conn_info->ssl);
-    if (ret <= 0)
+    bool connected = false;
+    bool should_retry = true;
+    int remaining_tries = MAX_CONNECT_RETRIES;
+    int ret = 1;
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libcfnet/tls_client.c]: Good catch. It was used as part of the `while` condition originally then I changed it to use more explicit `connected` instead.

File: libcfnet/tls_client.c [hunk 2]
```diff
@@ -310,8 +311,35 @@ int TLSTry(ConnectionInfo *conn_info)
...
+    while (!connected && should_retry)
+    {
+        ret = SSL_connect(conn_info->ssl);
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libcfnet/tls_client.c]: Might as well just add the assert in one extra commit.

**Pull request merged**